build-support/native-image/generate-native-image-for-pants-targets.bash

```diff
@@ -0,0 +1,205 @@
+    rm -rfv "$GENERATED_CONFIG_DIR"
+    mkdir -v "$GENERATED_CONFIG_DIR"
+    # TODO: jvm options are ignored from the command line, this is a bug and should be fixed.
+    export PANTS_COMPILE_ZINC_JVM_OPTIONS="+['-agentlib:native-image-agent=config-merge-dir=${GENERATED_CONFIG_DIR}']"
+    # NB: --worker-count=1 is because the `config-merge-dir` option to the native-image agent will
+    # clobber symbols from concurrent compiles.
+    MODE=debug ./pants -ldebug \
+            --no-zinc-native-image \
+            --resolve-coursier-capture-snapshots \
+            compile.zinc \
+            --execution-strategy=hermetic \
+            --worker-count=1 \
+            --no-incremental \
+            --no-use-classpath-jars \
+            --native-image-agent-library-dir="$(dirname "$agent_lib_path")" \
+            --cache-ignore \
+            "$@"
+    unset PANTS_COMPILE_ZINC_JVM_OPTIONS
+  fi >&2
+  normalize_path "$GENERATED_CONFIG_DIR"
+}
+
+# analyze tracing -- trim reflection down to just macros, then find those macros' targets with
+# classmap
+
+function trim_reflection_config_to_just_macros {
+  # Scala macros are the only entries in reflect-config.json that appear to matter for the purposes
+  # of compilation, so we filter down to only those.
+  # TODO: do compiler plugins look different?
+  jq '[.[] | select(.fields[]? | .name == "MODULE$")]' \
+     <"$GENERATED_CONFIG_JSON_FILE" \
+     >.tmp-reflect-config \
+    && mv -v \
+          .tmp-reflect-config \
+          "$GENERATED_CONFIG_JSON_FILE"
+}
+
+function extract_macro_target_names_from_pants {
+  ./pants classmap "$@" >.tmp-classmap || return "$?"
+
+  # Get the class names of all the macro usages, then filter down the above `./pants classmap` to
+  # find the targets which provide those classes.
+  jq -r '.[] | .name' <"$GENERATED_CONFIG_JSON_FILE" | while read -r class_name; do
+    grep -F "$class_name" .tmp-classmap \
+      | sed -Ee 's#^.* (.*)$#\1#g'
+  done | sort -u
+}
+
+# create a BUILD file of the appropriate dependencies
+
+function template_BUILD_file {
+  cat <<EOF
+# This file was generated by the scripts in https://github.com/pantsbuild/pants/tree/master/build-support/native-image/.
+# DO NOT EDIT!!!
+
+jvm_binary(
+  name='generated-native-image-binary',
+  main='org.pantsbuild.zinc.compiler.Main',
+  dependencies=[
+    $(sed -Ee 's#^(.*)$#"\1",#g')
+  ],
+)
+
+jvm_app(
+  name='native-image-deps',
+  basename='gen-native-image-basename',
+  binary=':generated-native-image-binary',
+  deployjar=True,
+)
+EOF
+}
+
+# generate the deps jar and reflect config at known locations!!!
+
+function generate_macro_deps_jar {
+  extract_macro_target_names_from_pants "$@" \
+    | if [[ -f "$GENERATED_BUILD_FILE" ]]; then
+    # If the generated BUILD file exists already, pull down buildozer to idempotently add the
+    # new deps!
+    _buildozer="$(download_buildozer)"
+    (xargs -t -P1 -L200 printf "//%s "; echo) | while read -r buildozer_deps_flattened; do
```

Comment:
This is pretty far removed from where its input comes from, and has pretty hard to follow control flow...

Reply:
This could benefit a lot from being split into about three functions with a few lines of comments, I think.

Reply:
I've split this into `chunk_and_normalize_pants_deps_for_buildozer`, `add_buildozer_macro_deps`, `generate_template_macro_deps_target`, `add_macro_deps_to_generated_target`, and `generate_macro_deps_jar`, with sufficient commenting to describe what's going on here.